### DISCUSSION TITLE
`no-param-reassign` to allow reassignment but disallow modifying properties

### DISCUSSION THREAD

**Discussion by GCSBOSS** [eslint/eslint#15769] [Ideas]
In my codebase I want to accept reassigning parameters, but disallow modifying properties of *the original objects coming from the outside*. That definition mught be a little sublte I guess. Here are some examples:

```js
function(a, b){
    a.foo = 'abc' // Disallowed since it changes the actual object coming from the arguments
    a = { foo: 'bar' } // Allowed to reassign
    a.foo = 'baz' // Allowed since this object does NOT come from the arguments

    let c = b;  // `c` now has an object coming from the arguments
    c.foo = 'bar' // Disallowed since you are modifying said object
}
```

Is this even viable to implement? Am I missing any caveats? Should this maybe be a new rule?

In case it's viable, is it possible to check in a nested level? 
```js
function(a){
    let b = { arg: a }; 
    b.arg.foo = 'bar' // Disallowed since you are modifying object comig from argument
}
```

Looking forward for everyone's comments.

**Comment by nzakas** [comment]
It’s possible. You should play with astexplorer.net to find the various patterns you’ll need to track.
